src/net_processing.cpp

```diff
@@ -947,6 +956,106 @@ void Misbehaving(NodeId pnode, int howmuch, const std::string& message) EXCLUSIV
+ * which provided such an object. This is used to determine whether to relay transactions to
+ * whitelisted peers, preventing us from relaying things which would result in them disconnecting
+ * us.
+ */
+static bool MayResultInDisconnect(const CValidationState& state, bool via_compact_block) {
+    switch (state.GetReason()) {
+    case ValidationInvalidReason::NONE:
+        return false;
+    // The node is providing invalid data:
+    case ValidationInvalidReason::CONSENSUS:
+    case ValidationInvalidReason::BLOCK_MUTATED:
+    case ValidationInvalidReason::CACHED_INVALID:
+        // Blocks (or headers) invalid for one of the above reasons may not be
+        // knowable to a high-bandwidth compact block peer, prior to relay.
+        // Headers that are invalid for reasons that should be known prior to
+        // block validation -- such as bad proof of work, too-early-time, or
+        // building off an invalid or missing block -- are punished regardless
+        // (see below).
+        return !via_compact_block;
+    case ValidationInvalidReason::BLOCK_INVALID_HEADER:
+    case ValidationInvalidReason::BLOCK_CHECKPOINT:
+    case ValidationInvalidReason::BLOCK_INVALID_PREV:
+    case ValidationInvalidReason::BLOCK_MISSING_PREV:
+        return true;
+    // Conflicting (but not necessarily invalid) data or different policy:
+    case ValidationInvalidReason::RECENT_CONSENSUS_CHANGE:
+    case ValidationInvalidReason::BLOCK_BAD_TIME:
+    case ValidationInvalidReason::TX_NOT_STANDARD:
+    case ValidationInvalidReason::TX_MISSING_INPUTS:
+    case ValidationInvalidReason::TX_WITNESS_MUTATED:
+    case ValidationInvalidReason::TX_CONFLICT:
+    case ValidationInvalidReason::TX_MEMPOOL_POLICY:
+        return false;
+    }
+    return false;
+}
+
+//! Returns true if the peer was punished (probably disconnected)
+static bool MaybePunishNode(NodeId nodeid, const CValidationState& state, bool via_compact_block, const std::string& message = "") {
+    switch (state.GetReason()) {
+    case ValidationInvalidReason::NONE:
+        break;
+    // The node is providing invalid data:
+    case ValidationInvalidReason::CONSENSUS:
+    case ValidationInvalidReason::BLOCK_MUTATED:
+        if (!via_compact_block) {
+            LOCK(cs_main);
+            Misbehaving(nodeid, 100, message);
+            return true;
+        }
+        break;
+    case ValidationInvalidReason::CACHED_INVALID:
+        {
+            LOCK(cs_main);
+            CNodeState *node_state = State(nodeid);
+            if (node_state == nullptr) {
+                break;
+            }
+
+            // Disconnect outbound (but not inbound) peers if on an invalid chain.
+            // Exempt HB compact block peers and manual connections.
+            if (!via_compact_block && !node_state->m_is_inbound && !node_state->m_is_manual_connection) {
+                Misbehaving(nodeid, 100, message);
```

Comment:
The comment says "disconnect", but the DoS score will also cause a ban here. Is that intentional? (it seems it's retaining existing behavior, so I assume it is).

Reply:
This is actually a behavior change from existing behavior, but hopefully a relatively harmless one.  Here's the relevant snippet from master:

https://github.com/bitcoin/bitcoin/blob/2945492424934fa360f86b116184ee8e34f19d0a/src/net_processing.cpp#L1552-L1585

It's a bit hard to decipher because of the multiple layers going on here, but basically `punish_duplicate_invalid` is only set to true for outbound and non-manual peers relaying us headers outside of HB compact block mode, and `nDoS` is set to 0 for the cached-invalid case:

https://github.com/bitcoin/bitcoin/blob/2945492424934fa360f86b116184ee8e34f19d0a/src/validation.cpp#L3341-L3357

So this does indeed result in a ban rather than just a disconnect for an outbound peer that announces an invalid header.  If there's no reason that this is materially worse than other ban behaviors that exist, then I think I'd prefer to stick with this behavior change for now, and try to improve ban-behavior globally after this PR (which should become much easier, now that banning is contained to one place in the code in a more understandable way).

I can update the comment though to make this clearer.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/15141#discussion_r255235104

Note: Thread pertains to commit "Fix handling of invalid headers" (9dd6fc18658b36b63b9f264676ac484879597b83)

Reply:
This appears to be correct to me. Obviously should update the comment to note this.

Reply:
Fixed comment in latest commit.